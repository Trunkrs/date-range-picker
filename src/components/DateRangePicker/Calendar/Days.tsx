import React from 'react'
import { getDaysInMonth } from 'date-fns'

import Day from '../Day'

export interface DaysProps {
  month: number
  year: number
  onDateClick?: (date: Date) => void
}

export const Days: React.FC<DaysProps> = ({ month, year, onDateClick }) => {
  const numberOfDays = getDaysInMonth(new Date(year, month))
  const daysArray = Array.from(Array(numberOfDays).keys())

  return (
    <>
      {daysArray.map((day) => {
        const display = day + 1
        return (
          <Day
            date={new Date(year, month, display)}
            key={`day${display}`}
            onClick={onDateClick}
          >
            {display}
          </Day>
        )
      })}
    </>
  )
}

export default React.memo(Days)
