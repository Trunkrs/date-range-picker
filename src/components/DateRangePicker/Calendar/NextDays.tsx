import React from 'react'
import Day from '../Day'

export interface DaysProps {
  month: number
  year: number
  showOutsideDays: boolean
}

export const NextDays: React.FC<DaysProps> = ({
  month,
  year,
  showOutsideDays
}) => {
  const nextDaysArray = Array.from(Array(14).keys())
  return (
    <React.Fragment>
      {nextDaysArray.map((day) => {
        const display = day + 1
        return (
          <Day
            date={new Date(year, month + 1, display)}
            key={`nextDay${display}`}
            show={showOutsideDays}
            outsideDayType='next'
            calendarMonth={new Date(year, month)}
          >
            {display}
          </Day>
        )
      })}
    </React.Fragment>
  )
}

export default React.memo(NextDays)
