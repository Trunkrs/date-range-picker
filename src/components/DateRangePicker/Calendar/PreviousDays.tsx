import React from 'react'

import useUtils from '../../../hooks/useUtils'
import Day from '../Day'

export interface DaysProps {
  month: number
  year: number
  showOutsideDays: boolean
}

export const PreviousDays: React.FC<DaysProps> = ({
  month,
  year,
  showOutsideDays
}) => {
  const { getDaysInMonth } = useUtils()

  const firstDayOfTheMonth = new Date(year, month, 1)
  const firstDayIndex = firstDayOfTheMonth.getDay()
  const prevMonthLastDay = getDaysInMonth(new Date(year, month - 1, 15))
  const startDay = prevMonthLastDay - firstDayIndex
  const prevDaysArray = Array.from(Array(prevMonthLastDay - startDay).keys())

  return (
    <>
      {prevDaysArray.map((day) => {
        const display = startDay + day + 1
        return (
          <Day
            date={new Date(year, month - 1, display)}
            key={`prevDay${display}`}
            show={showOutsideDays}
            outsideDayType='prev'
            calendarMonth={new Date(year, month)}
          >
            {display}
          </Day>
        )
      })}
    </>
  )
}

export default React.memo(PreviousDays)
