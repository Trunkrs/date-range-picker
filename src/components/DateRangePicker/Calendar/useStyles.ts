import createStyles from '@material-ui/core/styles/createStyles'
import makeStyles from '@material-ui/core/styles/makeStyles'
import { Theme } from '@material-ui/core/styles'

const styles = ({ palette, spacing }: Theme) =>
  createStyles({
    container: {
      backgroundColor: palette.common.white,
      display: 'inline-block',
    },
    calendar: {
      padding: spacing(2),
      display: 'flex',
      flexDirection: 'column',
    },
    month: {
      width: '100%',
      display: 'flex',
      justifyContent: 'center',
      padding: spacing(2),
      whiteSpace: 'nowrap',
    },
    spacer: {
      height: spacing(5),
    },
    dateGrid: {
      width: '100%',
      display: 'grid',
      gridTemplateColumns: `0px repeat(5, ${spacing(5)}px) 0px`,
      gridTemplateRows: `repeat(6, ${spacing(5)}px) 0px 0px 0px`,
    },
    showWeekends: {
      gridTemplateColumns: `repeat(7, ${spacing(5)}px) !important`,
    },
  })

export default makeStyles(styles)
