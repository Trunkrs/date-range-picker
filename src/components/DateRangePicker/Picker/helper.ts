import useUtils from '../../../hooks/useUtils'

export interface CalendarValue {
  date: Date
  key: number
}

export const getNextSetOfCalendarMonthsToView = (
  calendars: CalendarValue[],
  count = 1
): CalendarValue[] => {
  const { addMonths } = useUtils()

  const newCalendars = [...calendars]
  newCalendars.splice(0, 0, {
    date: addMonths(calendars[0].date, -count) as Date,
    key: calendars[0].key - count
  })
  newCalendars.pop()

  return newCalendars
}

export const getPreviousSetOfCalendarMonthToView = (
  calendars: CalendarValue[],
  count = 1
): CalendarValue[] => {
  const { addMonths } = useUtils()

  const newCalendars = [...calendars]
  const secondCalendar = calendars[calendars.length - 1]
  newCalendars.shift()
  newCalendars.push({
    date: addMonths(secondCalendar.date, count) as Date,
    key: secondCalendar.key + count
  })

  return newCalendars
}

export default getNextSetOfCalendarMonthsToView
