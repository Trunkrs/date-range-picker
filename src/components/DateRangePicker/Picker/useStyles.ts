import createStyles from '@material-ui/core/styles/createStyles'
import makeStyles from '@material-ui/core/styles/makeStyles'
import { Theme } from '@material-ui/core/styles'

const styles = ({ spacing }: Theme) =>
  createStyles({
    pickerRoot: {
      position: 'relative',
      overflow: 'hidden',
    },
    calendars: {
      display: 'flex',
      flexDirection: 'row',
      justifyContent: 'center',
    },
    controls: {
      top: spacing(3),
    },
    chevron: {
      width: spacing(4),
      height: spacing(4),
    },
    weekdays: {
      top: spacing(8),
    },
    fixedElement: {
      paddingLeft: spacing(2),
      paddingRight: spacing(2),
      position: 'absolute',
      width: '100%',
      display: 'flex',
      flexDirection: 'row',
      justifyContent: 'space-between',
    },
  })

export default makeStyles(styles)
