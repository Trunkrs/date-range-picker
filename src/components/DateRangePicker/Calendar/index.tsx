import React from 'react'

import clsx from 'clsx'

import useUtils from '../../../hooks/useUtils'
import Weekdays from '../Weekdays'
import Typography from '../../Typography'

import useStyles from './useStyles'
import Days from './Days'
import PreviousDays from './PreviousDays'
import NextDays from './NextDays'

export interface CalendarProps {
  month: number
  year: number
  className?: string
  showWeekends?: boolean
  showOutsideDays?: boolean
  hideWeekdayNames?: boolean
  onDateClick?: (date: Date) => void
}

export const Calendar = React.forwardRef<HTMLDivElement, CalendarProps>(
  (
    {
      month,
      year,
      className,
      showOutsideDays = false,
      showWeekends = false,
      hideWeekdayNames = false,
      onDateClick
    },
    ref
  ) => {
    const { formatByString } = useUtils()
    const classes = useStyles()

    return (
      <div ref={ref} className={clsx(classes.container, className)}>
        <div className={classes.calendar}>
          <div className={classes.month}>
            <Typography variant='textBold'>
              {formatByString(new Date(year, month), 'MMMM yyyy')}
            </Typography>
          </div>

          {hideWeekdayNames ? (
            <div className={classes.spacer} />
          ) : (
            <Weekdays showWeekends={showWeekends} />
          )}

          <div
            className={clsx(
              classes.dateGrid,
              showWeekends && classes.showWeekends
            )}
          >
            <PreviousDays
              month={month}
              year={year}
              showOutsideDays={showOutsideDays}
            />
            <Days month={month} year={year} onDateClick={onDateClick} />
            <NextDays
              month={month}
              year={year}
              showOutsideDays={showOutsideDays}
            />
          </div>
        </div>
      </div>
    )
  }
)

export default React.memo(Calendar)
