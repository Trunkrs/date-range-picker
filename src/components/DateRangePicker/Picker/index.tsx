import React, {
  useState,
  useCallback,
  useMemo,
  useContext,
  useEffect,
} from 'react'

import clsx from 'clsx'
import { useTheme } from '@material-ui/core/styles'

import { useTransition, config, animated } from 'react-spring'

import addMonths from 'date-fns/addMonths'
import subMonths from 'date-fns/subMonths'
import isSameMonth from 'date-fns/isSameMonth'
import differenceInMonths from 'date-fns/differenceInMonths'
import ButtonBase from '@material-ui/core/ButtonBase'
import ChevronLeftIcon from '@material-ui/icons/ChevronLeft'
import ChevronRightIcon from '@material-ui/icons/ChevronRight'

import Calendar from '../Calendar'
import Weekdays from '../Weekdays'
import useStyles from './useStyles'
import useRefValue from '../../../hooks/useRefValue'
import DateRange from '../DateRange'
import {
  getNextSetOfCalendarMonthsToView,
  getPreviousSetOfCalendarMonthToView,
} from './helper'

export interface PickerProps {
  initialDate: Date
  showWeekends?: boolean
  showOutsideDays?: boolean
  onDateClick?: (date: Date) => void
}

export const Picker: React.FC<PickerProps> = ({
  initialDate,
  showWeekends = true,
  showOutsideDays = true,
  onDateClick,
}) => {
  const [calendars, setCalendars] = useState([
    { date: subMonths(initialDate, 1), key: -1 },
    { date: initialDate, key: 0 },
    { date: addMonths(initialDate, 1), key: 1 },
    { date: addMonths(initialDate, 2), key: 2 },
  ])
  const [adjustment, setAdjustment] = useState(0)
  const dateRange = useContext(DateRange)
  const prevDate1 = useRefValue(dateRange.dateStart)

  const theme = useTheme()
  const classes = useStyles()

  const width = useMemo(() => {
    return showWeekends ? theme.spacing(39) : theme.spacing(29)
  }, [theme, showWeekends])

  const handlePrev = useCallback(
    (_e?: React.SyntheticEvent, count = 1) => {
      const newCalendars = getNextSetOfCalendarMonthsToView(calendars, count)
      setCalendars(newCalendars)
    },
    [calendars],
  )

  const handleNext = useCallback(
    (_e?: React.SyntheticEvent, count = 1) => {
      const newCalendars = getPreviousSetOfCalendarMonthToView(calendars, count)
      setCalendars(newCalendars)
    },
    [calendars],
  )

  const handleWheel = useCallback(
    (event: React.WheelEvent) => {
      if (event.deltaY > 0) {
        handleNext()
        return
      }

      handlePrev()
    },
    [handleNext, handlePrev],
  )

  // animate to the correct calendar if date1 is out of view
  useEffect(() => {
    if (!dateRange.dateStart || dateRange.dateStart === prevDate1) return

    const isDateOutOfView =
      !isSameMonth(dateRange.dateStart, calendars[1].date) ||
      !isSameMonth(dateRange.dateStart, calendars[2].date)

    if (isDateOutOfView) {
      const diff = differenceInMonths(dateRange.dateStart, calendars[1].date)
      const skip = Math.abs(diff) - 5

      if (diff > 5) {
        handleNext(undefined, skip)
        setAdjustment(5)
        return
      }

      if (diff < -5) {
        handlePrev(undefined, skip)
        setAdjustment(-5)
        return
      }

      setAdjustment(diff)
    }
  }, [dateRange, calendars, adjustment, prevDate1, handlePrev, handleNext])

  // animate to the correct calendar until there is no more adjustment needed
  useEffect(() => {
    if (adjustment === 0) return

    if (adjustment < 0) {
      handlePrev()
      setAdjustment(adjustment + 1)
      return
    }

    handleNext()
    setAdjustment(adjustment - 1)
  }, [adjustment, handleNext, handlePrev])

  const transition = useTransition(calendars, {
    key: (calendar: any) => calendar.key,
    initial: { width: '100%' },
    from: { width: '0%' },
    enter: { width: '100%' },
    leave: { width: '0%' },
    config: {
      ...config.stiff,
      clamp: true,
    },
  })

  return (
    <div
      className={classes.pickerRoot}
      style={{ width: width * 2 }}
      onWheel={handleWheel}
    >
      <div className={classes.calendars}>
        {transition((value, calendar) => (
          <animated.div style={value}>
            <Calendar
              month={calendar.date.getMonth()}
              year={calendar.date.getFullYear()}
              showWeekends={showWeekends}
              hideWeekdayNames
              showOutsideDays={showOutsideDays}
              onDateClick={onDateClick}
            />
          </animated.div>
        ))}
      </div>

      <div className={clsx(classes.controls, classes.fixedElement)}>
        <ButtonBase onClick={handlePrev}>
          <ChevronLeftIcon className={classes.chevron} />
        </ButtonBase>
        <ButtonBase onClick={handleNext}>
          <ChevronRightIcon className={classes.chevron} />
        </ButtonBase>
      </div>

      <div className={clsx(classes.weekdays, classes.fixedElement)}>
        <Weekdays showWeekends={showWeekends} />
        <Weekdays showWeekends={showWeekends} />
      </div>
    </div>
  )
}

export default Picker
