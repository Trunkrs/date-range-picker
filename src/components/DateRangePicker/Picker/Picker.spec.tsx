import React from 'react'

import { Globals } from 'react-spring'
import { createShallow } from '@material-ui/core/test-utils'

import Picker, { PickerProps } from '.'
import Calendar from '../Calendar'

jest.unmock('react-spring')
Globals.assign({
  skipAnimation: true,
})

const createProps = (merge: Partial<PickerProps> = {}): PickerProps =>
  ({
    initialDate: new Date(2020, 9),
    onDateClick: jest.fn(),
    ...merge,
  } as PickerProps)

describe('<Picker />', () => {
  const shallow = createShallow()

  it('should match snapshot', () => {
    const wrapper = shallow(
      <Picker {...createProps({ showWeekends: false })} />,
    )
    expect(wrapper).toMatchSnapshot()
  })

  it('should navigate to next calendar(month) on mouse wheel', (done) => {
    const wrapper = shallow(<Picker {...createProps()} />)
    wrapper.simulate('wheel', { deltaY: 1 })
    wrapper.update()

    setTimeout(() => {
      expect(wrapper.find(Calendar).at(1).prop('month')).toBe(10)
      done()
    }, 100)
  })

  it('should navigate to previous calendar(month) on mouse wheel', (done) => {
    const wrapper = shallow(<Picker {...createProps()} />)
    wrapper.simulate('wheel', { deltaY: -1 })
    wrapper.update()

    setTimeout(() => {
      expect(wrapper.find(Calendar).at(1).prop('month')).toBe(8)
      done()
    }, 100)
  })
})
