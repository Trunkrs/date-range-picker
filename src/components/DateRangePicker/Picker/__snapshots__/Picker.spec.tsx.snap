// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Picker /> should match snapshot 1`] = `
<div
  className="makeStyles-pickerRoot-1"
  onWheel={[Function]}
  style={
    Object {
      "width": 464,
    }
  }
>
  <div
    className="makeStyles-calendars-2"
  >
    <Animated(div)
      key="-1"
      style={
        Object {
          "width": "100%",
        }
      }
    >
      <Memo(ForwardRef)
        hideWeekdayNames={true}
        month={8}
        onDateClick={[MockFunction]}
        showOutsideDays={true}
        showWeekends={false}
        year={2020}
      />
    </Animated(div)>
    <Animated(div)
      key="0"
      style={
        Object {
          "width": "100%",
        }
      }
    >
      <Memo(ForwardRef)
        hideWeekdayNames={true}
        month={9}
        onDateClick={[MockFunction]}
        showOutsideDays={true}
        showWeekends={false}
        year={2020}
      />
    </Animated(div)>
    <Animated(div)
      key="1"
      style={
        Object {
          "width": "100%",
        }
      }
    >
      <Memo(ForwardRef)
        hideWeekdayNames={true}
        month={10}
        onDateClick={[MockFunction]}
        showOutsideDays={true}
        showWeekends={false}
        year={2020}
      />
    </Animated(div)>
    <Animated(div)
      key="2"
      style={
        Object {
          "width": "100%",
        }
      }
    >
      <Memo(ForwardRef)
        hideWeekdayNames={true}
        month={11}
        onDateClick={[MockFunction]}
        showOutsideDays={true}
        showWeekends={false}
        year={2020}
      />
    </Animated(div)>
  </div>
  <div
    className="makeStyles-controls-3 makeStyles-fixedElement-6"
  >
    <WithStyles(ForwardRef(ButtonBase))
      onClick={[Function]}
    >
      <Memo(ForwardRef(ChevronLeftIcon))
        className="makeStyles-chevron-4"
      />
    </WithStyles(ForwardRef(ButtonBase))>
    <WithStyles(ForwardRef(ButtonBase))
      onClick={[Function]}
    >
      <Memo(ForwardRef(ChevronRightIcon))
        className="makeStyles-chevron-4"
      />
    </WithStyles(ForwardRef(ButtonBase))>
  </div>
  <div
    className="makeStyles-weekdays-5 makeStyles-fixedElement-6"
  >
    <Memo(Weekdays)
      showWeekends={false}
    />
    <Memo(Weekdays)
      showWeekends={false}
    />
  </div>
</div>
`;
